gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/vertx/web/handler/impl/internal/mfa/MFAEnrollStep.java

```diff
@@ -47,14 +57,119 @@ public MFAEnrollStep(Handler<RoutingContext> wrapper, RuleEngine ruleEngine, Fac
+                flow.exit(this);
+                return;
+            } else if (isChallengeConditional(client) && challengeConditionNotSatisfied(client, filterContext, routingContext)) {
+                flow.exit(this);
+                return;
+            } else if (isChallengeRiskBased(client) && challengeConditionNotSatisfied(client, filterContext, routingContext)) {
+                flow.exit(this);
+                return;
+            }
+        }
+
+        flow.doNext(routingContext);
+    }
+
+
+    private boolean isEnrollOptional(Client client) {
+        final EnrollSettings settings = MfaUtils.getEnrollSettings(client);
+        return MfaEnrollType.OPTIONAL == settings.getType();
+    }
+
+    private boolean isEnrollRequired(Client client) {
+        final EnrollSettings settings = MfaUtils.getEnrollSettings(client);
+        return MfaEnrollType.REQUIRED == settings.getType();
+    }
+
+    private boolean isEnrollConditional(Client client) {
+        final EnrollSettings settings = MfaUtils.getEnrollSettings(client);
+        return MfaEnrollType.CONDITIONAL == settings.getType();
+    }
+
+    private boolean enrollConditionNotSatisfied(Client client, MfaFilterContext context, RoutingContext routingContext) {
+        final EnrollSettings settings = MfaUtils.getEnrollSettings(client);
+        Boolean result = ruleEngine.evaluate(Optional.ofNullable(settings.getEnrollmentRule()).orElse("{}"),
+                context.getEvaluableContext(), Boolean.class, false);
+        routingContext.session().put(MFA_ENROLLMENT_CONDITION_SATISFIED, result);
+        return !result;
+    }
+
+    private boolean isRisk(Client client, MfaFilterContext context) {
+        if (hasAdaptiveAuthAuthRule(client)) {
+            final String adaptiveAuthRule = MfaUtils.getAdaptiveMfaStepUpRule(client);
+            return ruleEngine.evaluate(adaptiveAuthRule, context.getEvaluableContext(), Boolean.class, false);
         }
+
+        //todo need to check this
+        return false;
+    }
+
+    private boolean isChallengeRequired(Client client) {
+        ChallengeSettings settings = MfaUtils.getChallengeSettings(client);
+        return settings.getType() == MfaChallengeType.REQUIRED;
+    }
+
+    private boolean isChallengeConditional(Client client) {
+        ChallengeSettings settings = MfaUtils.getChallengeSettings(client);
+        return settings.getType() == MfaChallengeType.CONDITIONAL;
+    }
+
+    private boolean isChallengeRiskBased(Client client) {
+        ChallengeSettings settings = MfaUtils.getChallengeSettings(client);
+        return settings.getType() == MfaChallengeType.RISK_BASED;
+    }
+
+    private boolean challengeConditionNotSatisfied(Client client, MfaFilterContext context, RoutingContext routingContext) {
+        final ChallengeSettings settings = MfaUtils.getChallengeSettings(client);
+        Boolean result = ruleEngine.evaluate(Optional.ofNullable(settings.getChallengeRule()).orElse("{}"),
```

Comment:
> should not be? .orElse("")

instead of empty json object or empty string, we should provide a "false" string to explicit the expected default value